notes.lua

```diff
@@ -0,0 +1,322 @@
+                    view_id='buttons',
+                    frame={b=0,h=3},
+                    autoarrange_subviews=true,
+                    subviews={
+                        edit_mode and widgets.TextButton{
+                            view_id='Save',
+                            frame={h=1},
+                            label='Save',
+                            key='CUSTOM_ALT_S',
+                            on_activate=function() self:saveNote() end,
+                            enabled=function() return #self.subviews.name:getText() > 0 end,
+                        } or widgets.TextButton{
+                            view_id='Create',
+                            frame={h=1},
+                            label='Create',
+                            key='CUSTOM_ALT_S',
+                            on_activate=function() self:createNote() end,
+                            enabled=function() return #self.subviews.name:getText() > 0 end,
+                        },
+                        edit_mode and widgets.TextButton{
+                            view_id='delete',
+                            frame={h=1},
+                            label='Delete',
+                            key='CUSTOM_ALT_D',
+                            on_activate=function() self:deleteNote() end,
+                        } or nil,
+                    }
+                }
+            },
+        },
+    }
+end
+
+function NoteManager:createNote()
+    local x, y, z = pos2xyz(df.global.cursor)
+    if x == nil then
+        print('Enable keyboard cursor to add a note.')
+        return
+    end
+
+    local name = self.subviews.name:getText()
+    local comment = self.subviews.comment:getText()
+
+    if #name == 0 then
+        print('Note need at least a name')
+        return
+    end
+
+    local waypoints = df.global.plotinfo.waypoints
+    local notes = df.global.plotinfo.waypoints.points
+
+    notes:insert("#", {
+        new=true,
+
+        id = waypoints.next_point_id,
+        tile=88,
+        fg_color=7,
+        bg_color=0,
+        name=name,
+        comment=comment,
+        pos=xyz2pos(x, y, z)
+    })
+    waypoints.next_point_id = waypoints.next_point_id + 1
+
+    if self.on_update then
+        self.on_update()
+    end
+
+    self:dismiss()
+end
+
+function NoteManager:saveNote()
+    if self.note == nil then
+        return
+    end
+
+    local name = self.subviews.name:getText()
+    local comment = self.subviews.comment:getText()
+
+    if #name == 0 then
+        print('Note need at least a name')
+        return
+    end
+
+    local notes = df.global.plotinfo.waypoints.points
+    self.note.name = name
+    self.note.comment = comment
+
+    if self.on_update then
+        self.on_update()
+    end
+
+    self:dismiss()
+end
+
+function NoteManager:deleteNote()
+    if self.note == nil then
+        return
+    end
+
+    for ind, note in pairs(df.global.plotinfo.waypoints.points) do
+        if note == self.note then
```

Comment:
be careful about comparing pointers. it doesn't always behave the way you expect (e.g. when used as keys in maps). For tests like these, the common practice in DF/DFHack is to compare the IDs, not the pointers. `if note.id == self.note.id`